### PR TITLE
Add TwistStamped support via TwistPublisher and TwistSubscriber

### PR DESCRIPTION
**Pull request by Ryanf55**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1594  |
| Primary OS tested on | Ubuntu 22.04 |
| Robotic platform tested on | CI |

---

## Description of contribution in a few bullet points

* Add a new util class for publishing either Twist or TwistStamped called `TwistPublisher`
* Add a new util class for subscribing to either Twist or TwistStamped called `TwistSubscriber`
* Add a new parameter for selecting to stamp the twist data to each util
* Consume TwistPublisher in nav2_controller, among other nodes
* Consume TwistSubscriber in nav2_velocityy_smoother, among other nodes
* Implements the approach described [here](https://github.com/ros-planning/navigation2/issues/1594#issuecomment-1675262038): `But, I suppose we could make a velocity publisher wrapper in nav2_util that deals with that parameterization outside of the application code (and in a way we can remove the non-stamped support more gradually)`.

## Description of documentation updates required from your changes

* Would need to document the new optional parameter behavior in the normal docs on top of the README.

## Future work that may be required in bullet points

* Related PR's in gazebo, etc, to convert the ecosystem to TwistStamped

## Test Instructions

```bash
colcon test --packages-select nav2_util --event-handlers=console_cohesion+ --ctest-args " -R" " test_twist"
```

Quick iteration
```bash
colcon build --packages-select nav2_util && colcon test --packages-select nav2_util --event-handlers=console_cohesion+ --ctest-args " -R" " test_twist"
```

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (Ryanf55) [inline nav2_util/include/nav2_util/twist_publisher.hpp:45]: copy-pasta errror

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (Ryanf55) [inline nav2_util/include/nav2_util/twist_publisher.hpp:102]: remove

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/twist_publisher.hpp:118]: This doesn't need to be stored

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/twist_publisher.hpp:98]: A non-stamped version that adds a stamp if `is_stamped_` is true? Basically the counter-implementation of this one

**Review by SteveMacenski — COMMENTED** [review]
Generally LGTM other than missing test coverage + documentation updates on the new param / feature

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (SteveMacenski) [reply nav2_util/include/nav2_util/twist_publisher.hpp:118]: There are various reasons why storing the shared pointer creates problems

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_publisher.hpp]: I just removed it, was using it during testing.

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_publisher.hpp:97]: I was thinking about this, and was planning to add it when it's needed, however it seems a lot of NAV2 internals are using the stamped version anyways.

Comment (SteveMacenski) [reply nav2_util/include/nav2_util/twist_publisher.hpp:97]: Oh ok, then I don't think its strictly necessary then.

The only thing I'd say in counter to that is that if others want to use this to stamp their twists (with a param/method arg for the base_frame) that could be a nice add on. But given your expertise / applications i think you're in the best position to tell me if that's even something desired. If not, feel free to ignore.

**Comment by SteveMacenski** [comment]
I think just unit testing is the main blocker here

**Comment by SteveMacenski** [comment]
Add unit tests for this new object to make sure its working properly e.g. 
- Check when you ask it to pub with stamp it does
- Check when you ask it to pub with non-stamp it does
- Check the different APIs you've created work / don't crash when called correctly
- Check that get sub count works properly in getting the right stamp vs non-stamp

^ the last 2 are obvious, but we don't know how its going to be modified over time. Also, its dereferencing pointers that will crash if the logic is ever wrong since the other publisher isn't allocated. So exercising the code in CI promises it won't crash applications later. 

Just basic tests - nothing nuts

**Comment by Ryanf55** [comment]
> Add unit tests for this new object to make sure its working properly e.g.
> 
> * Check when you ask it to pub with stamp it does
> * Check when you ask it to pub with non-stamp it does
> * Check the different APIs you've created work / don't crash when called correctly
> * Check that get sub count works properly in getting the right stamp vs non-stamp
> 
> ^ the last 2 are obvious, but we don't know how its going to be modified over time. Also, its dereferencing pointers that will crash if the logic is ever wrong since the other publisher isn't allocated. So exercising the code in CI promises it won't crash applications later.
> 
> Just basic tests with gtest like the other utils - nothing nuts.

Sounds good, can do!

File: nav2_util/test/test_twist_publisher.cpp
Comment (SteveMacenski) [inline nav2_util/test/test_twist_publisher.cpp:1]: Update

File: nav2_util/test/test_twist_publisher.cpp
Comment (SteveMacenski) [inline nav2_util/test/test_twist_publisher.cpp:49]: This is insufficient. You need to have something that accepts that message to make sure the action is actually taken. Make a subscription and check that it was received / the values. 

Also, Send it both stamped/unstamped and also make sure both are received and the non-stamped was properly stamped. And check that the subscription count is incremented with the subscriber is added. 

Also a second test like this for `enable_stamped_cmd_vel = true` . 

Basically, cover the code :-)

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_util/test/test_twist_publisher.cpp
Comment (Ryanf55) [reply nav2_util/test/test_twist_publisher.cpp]: Done in 169f7b3778d3198c8266ba7ec026ae5fc66a58bc

File: nav2_util/test/test_twist_publisher.cpp
Comment (Ryanf55) [reply nav2_util/test/test_twist_publisher.cpp]: Done in 2bc7d9dd2fa0c1a43368927788ad6b2cd1e1f233.

I added a subscriber, and cover both the default value of the param, as well as when the param is modified to enforce stamping. The subscription count is now tested, as well as the activation state and all the transitions.

**Comment by Ryanf55** [comment]
I've added much more comprehensive testing. Currently, CI is failing a test:
https://app.circleci.com/pipelines/github/ros-planning/navigation2/9938/workflows/f5cb1d0d-0c04-48cb-9a51-1a7771b7b613/jobs/31717?invite=true#step-110-9217_47
```
    [ RUN      ] UtilsTests.SmootherTest
    /opt/overlay_ws/src/navigation2/nav2_mppi_controller/test/utils_test.cpp:381: Failure
    Expected: (smoothed_val) < (original_val), actual: 6.3252e+25 vs 11.9772
    [  FAILED  ] UtilsTests.SmootherTest (0 ms)
```

I don't think my PR causes this failure - that test has no reliance on headers I changed. 

As far as all the other uses of Twist in NAV2, I;ve learned some nodes subscribe and republish Twist such as the velocity smoother. In order to have this backward-compatible support for TwistStamped in the entire stack, I expect I'll need to create a mirrored `TwistSubscriber` just like the `TwistPublisher`.

**Comment by SteveMacenski** [comment]
I think this PR so that we have a clean migration over to it. Just merging this alone could break users that enable that param and don't have the ability to enable that new param with the velocity smoother and collision monitor. 

I looked over the codecov results and the code and this is good to me now, so consider the publisher approved!

**Comment by Ryanf55** [comment]
I was curious if you had any ideas on the TwistStamped subscribers how to deal with the callbacks needing to have a certain signature. For example: `AssistedTeleop::teleopVelocityCallback`

My thought is to add overrides for all node callbacks to also work with TwistStamped, and convert all internal variables to TwistStamped typed. Then, once the ecosystem (gazebo, etc) is moved over to TwistStamped, the pure Twist callbacks can be deprected on `rolling` and removed in J-turtle.

**Comment by SteveMacenski** [comment]
> I was curious if you had any ideas on the TwistStamped subscribers how to deal with the callbacks needing to have a certain signature. For example: AssistedTeleop::teleopVelocityCallback

That sample is pretty easy: the twist subscriber has some `getTwist` method for when it needs to be gotten and otherwise can just be a subobject of this behavior. There may also need to be a `reset()` to set the twist to default (or just have a bool for `received_` that gets reset to mark validness).

Others may be more complex. You may want to enumerate them and we can discuss them individually so you don't waste time

> My thought is to add overrides for all node callbacks to also work with TwistStamped, and convert all internal variables to TwistStamped typed. Then, once the ecosystem (gazebo, etc) is moved over to TwistStamped, the pure Twist callbacks can be deprected on rolling and removed in J-turtle.

I don't fully follow. Like have both a stamped and non-stamped callbacks, each registered, and the stamped just passes the `msg.twist` to the non-stamped version that implements the logic? Or only actually `create_subscription` based on the twist stamped-ness so only one's active on the network

**Comment by Ryanf55** [comment]
> There may also need to be a

Goal: Only subscribe with either Twist or TwistStamped, not both to eliminate overhead.

Based on your feedback, it's leading me to have the TwistSubscriber class constructor take in both signatures (Twist and TwistStamped) of the subscriber callback. It will bind whichever the node is configured for. The problem with that approach, assuming the callbacks are overrides is that the template arg deduction in bind fails. 
```
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp: In member function ‘virtual nav2_util::CallbackReturn nav2_collision_monitor::CollisionMonitor::on_configure(const rclcpp_lifecycle::State&)’:
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:68:14: error: no matching function for call to ‘bind(<unresolved overloaded function type>, nav2_collision_monitor::CollisionMonitor*, const std::_Placeholder<1>&)’
   68 |     std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/11/pstl/glue_algorithm_defs.h:13,
                 from /usr/include/c++/11/algorithm:74,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executor.hpp:18,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors/multi_threaded_executor.hpp:25,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors.hpp:21,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/rclcpp.hpp:167,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp:22,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:15:
/usr/include/c++/11/functional:789:5: note: candidate: ‘template<class _Func, class ... _BoundArgs> typename std::_Bind_helper<std::__is_socketlike<_Func>::value, _Func, _BoundArgs ...>::type std::bind(_Func&&, _BoundArgs&& ...)’
  789 |     bind(_Func&& __f, _BoundArgs&&... __args)
      |     ^~~~
/usr/include/c++/11/functional:789:5: note:   template argument deduction/substitution failed:
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:68:14: note:   couldn’t deduce template parameter ‘_Func’
   68 |     std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/11/pstl/glue_algorithm_defs.h:13,
                 from /usr/include/c++/11/algorithm:74,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executor.hpp:18,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors/multi_threaded_executor.hpp:25,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/executors.hpp:21,
                 from /opt/ros/rolling/include/rclcpp/rclcpp/rclcpp.hpp:167,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/include/nav2_collision_monitor/collision_monitor_node.hpp:22,
                 from /home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:15:
/usr/include/c++/11/functional:813:5: note: candidate: ‘template<class _Result, class _Func, class ... _BoundArgs> typename std::_Bindres_helper<_Result, _Func, _BoundArgs>::type std::bind(_Func&&, _BoundArgs&& ...)’
  813 |     bind(_Func&& __f, _BoundArgs&&... __args)
      |     ^~~~
/usr/include/c++/11/functional:813:5: note:   template argument deduction/substitution failed:
/home/ryan/Dev/ros2_ws/src/navigation2/nav2_collision_monitor/src/collision_monitor_node.cpp:68:14: note:   couldn’t deduce template parameter ‘_Result’
   68 |     std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [CMakeFiles/collision_monitor_core.dir/build.make:76: CMakeFiles/collision_monitor_core.dir/src/collision_monitor_node.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:186: CMakeFiles/collision_monitor_core.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< nav2_collision_monitor [2.51s, exited with code 2]
```

For now, I'll just name the callbacks differently (easy) or switch to lambdas (harder).

**Comment by Ryanf55** [comment]
Hi @SteveMacenski,

I've gotten all the features working. I'd like to start iterating on getting to 100% code coverage, or at least no decrease in code coverage such that you will be happy merging. 

I know that CI runs a tool that shows you a diff of your PR against the target branch, colorizing with respect to changes in code coverage. Is there any way to use it locally so I can iterate faster than waiting for the build and test workflow in CI to run?

**Comment by Ryanf55** [comment]
Looks like the latest build has flake8 failures in CI are totally unrelated to my changes. I do not know what I should do to resolve those:
https://app.circleci.com/pipelines/github/ros-planning/navigation2/10324/workflows/e2990a21-f6d6-428b-8b0c-8029562638b1/jobs/32684?invite=true#step-110-7208_26

File: nav2_behaviors/include/nav2_behaviors/plugins/drive_on_heading.hpp
Comment (SteveMacenski) [inline nav2_behaviors/include/nav2_behaviors/plugins/drive_on_heading.hpp:144]: Go back to the unique pointer + move. It removes jitter

File: nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp
Comment (SteveMacenski) [inline nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp:300]: Commented otu?

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/twist_publisher.hpp:63]: Config guide needs this for all nodes that use it to know about this option. Migration guide too

File: nav2_util/include/nav2_util/twist_subscriber.hpp
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/twist_subscriber.hpp:82]: Q: why not have a callback here and a `getTwist` & `getTwistStamped` member?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Overall, LGTM. One thing I'd ask is to review the README files of changed packages + Nav2 documentation to update the use of Twist or TwistStamped. I know its annoying to do docs, but you should be able to get through skimming each of the pages and updating that within 15 minutes and would do a world to help users down the line if they didn't know about this to have hints in the docs easily

File: nav2_behaviors/include/nav2_behaviors/plugins/drive_on_heading.hpp
Comment (SteveMacenski) [reply nav2_behaviors/include/nav2_behaviors/plugins/drive_on_heading.hpp:144]: ^ and for all cases like this one

File: nav2_util/include/nav2_util/twist_subscriber.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_subscriber.hpp:82]: I didn't know what the behavior of getTwistStamped would be if the node received Twist. Does it re-stamp the timestamp automagically? Right now, there is no made up timestamps. Either it has timestamp data, or it doesn't and you aren't allowed to use it.

Comment (SteveMacenski) [reply nav2_util/include/nav2_util/twist_subscriber.hpp:82]: Could throw?

I'm not necessarily asking for a change, just a question to see if it would be sensible to make this similar the `OdomSmoother` or other subscriber-data objects which hold their data and get when requested. I suppose often the Twist Subscription clients are those that need event-based access to the twist data, not polling based, which would be a reasonable answer for why this method is necessary as well.

**Comment by SteveMacenski** [comment]
I fixed the merge conflict

**Comment by SteveMacenski** [comment]
Hi,

I think my comments basically were the end of the road of changes needed before merge. If you wanted this in before Thanksgiving, that's definitely on the table if updates are made :-)

**Comment by Ryanf55** [comment]
> Hi,
> 
> I think my comments basically were the end of the road of changes needed before merge. If you wanted this in before Thanksgiving, that's definitely on the table if updates are made :-)

Thanks for the time. I'm swamped this week, but can get those last few things done next week.

File: nav2_util/include/nav2_util/twist_subscriber.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_subscriber.hpp:82]: > Could throw?
> 
> I'm not necessarily asking for a change, just a question to see if it would be sensible to make this similar the `OdomSmoother` or other subscriber-data objects which hold their data and get when requested. I suppose often the Twist Subscription clients are those that need event-based access to the twist data, not polling based, which would be a reasonable answer for why this method is necessary as well. So perhaps its just not sensible?

I think it could be sensible, but I worry that it is a bit potential higher impact (probability of bugs). Without sufficient test infrastructure to run the whole system tests with a general parameter override globally for the new param, I'm hesitant to introduce data-sharing modifications.

File: nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp
Comment (Ryanf55) [reply nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp:300]: Using std::move on the command results in a failure to compile.  It was not clear why TwistStamped needs to be a unique pointer, so I just pass it by value into publish.

Comment (Ryanf55) [reply nav2_behaviors/include/nav2_behaviors/timed_behavior.hpp]: Here's a reference I found that says the usage I switched to is also valid: 
https://github.com/ros2/rclcpp/pull/676#issuecomment-489381722

What's not clear is why std::move is broken, unless the `rclcpp` maintainers changed the API again since that update.

**Comment by SteveMacenski** [comment]
You should see things pass properly. That looks like just a one-off flaky test. What is your issues locally?

File: nav2_behaviors/include/nav2_behaviors/plugins/drive_on_heading.hpp
Comment (SteveMacenski) [reply nav2_behaviors/include/nav2_behaviors/plugins/drive_on_heading.hpp:144]: I don't know what you're talking about with new syntax. This is Rolling and we have every other publisher using `std::move` on a unique pointer. Please update all publications to use publish moved semantics unique pointers. See pretty much every other publisher in the stack for an example :laughing:

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (SteveMacenski) [reply nav2_util/include/nav2_util/twist_publisher.hpp:63]: This is still important before a merge. It should be as easy as adding this parameter to each of the configuration guide pages + the YAML at the bottom -- then writing up a short blurb in the migration guide about it

File: nav2_velocity_smoother/README.md
Comment (SteveMacenski) [inline nav2_velocity_smoother/README.md:62]: Add a comment here about what this represents like the others

File: nav2_util/README.md
Comment (SteveMacenski) [inline nav2_util/README.md:34]: You can remove the usage subsection in entirety

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/twist_publisher.hpp:58]: `)` on line above, this looks like a strange auto-reformatting choice

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/twist_publisher.hpp:72]: Line extension really needed?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
Any update here?

File: nav2_velocity_smoother/src/velocity_smoother.cpp
Comment (SteveMacenski) [inline nav2_velocity_smoother/src/velocity_smoother.cpp:131]: This server needs to publish out a TwistStamped, else this PR is essentially not valuable since the pipeline goes from the controller server -> collision monitor -> velocity smoother -> robot base / drone. The outputs of the velocity smoother + collision monitor also need to be twiststamped

Comment (SteveMacenski) [reply nav2_velocity_smoother/src/velocity_smoother.cpp:131]: I think that necessitates the inversion of the changes in this file. The TwistStamped should be the actual callback implementing the actual behavior (e.g. storing header to publish out). The un-stamped Cb should create a header to meet the API to go into the stamped callback to use the rest of the pipeline & publish back out unstamped

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:66]: Same with velocity smoother: the output here also needs to be stamped and probably changes the way that this should be handled.

Here and in the velocity smoother: why not use the specific constructors for the specific stamped/unstamped parameter so we're not registering both cbs when unnecessary?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
All of the open requests from the last review are also still relevant (but all small). I'll handle the re-pointer-ification once everything else is handled. 

The least of it, but CI is complaining about a small number of linting issues. Usual bits on configuration guide for new parameter + migration guide for capability documentation

**Comment by SteveMacenski** [comment]
I handled the conflict

File: nav2_util/include/nav2_util/twist_subscriber.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_subscriber.hpp:89]: I've implemented the support for the `Thought 2` with a second constructor.

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_publisher.hpp:72]: This is what ament_uncrustify wants. I just reran it.

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_publisher.hpp]: Interesting. ament_uncrustify accepts code in either configuration as valid. Sorry, I'm used to clang-tidy which always results in the same output regardless of what the input code is.

Fixed!

Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_publisher.hpp]: e65fed5e66dac9f0cc264766c078b8b31332f1cd

File: nav2_util/README.md
Comment (Ryanf55) [reply nav2_util/README.md]: d074e000c6d6130d0ba6d35b8f19c0af685082ce

File: nav2_velocity_smoother/README.md
Comment (Ryanf55) [reply nav2_velocity_smoother/README.md]: b94181e0b8d6a393d49f3b8e56d8519e8e72c21b

File: nav2_velocity_smoother/src/velocity_smoother.cpp
Comment (Ryanf55) [reply nav2_velocity_smoother/src/velocity_smoother.cpp:131]: Done. I also switched it to a unique pointer instead of a shared pointer. 
6b110684405a58ed8a906acd1da7a3aef58af171

File: nav2_collision_monitor/src/collision_monitor_node.cpp
Comment (Ryanf55) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:66]: Yep, good spot. I'll handle both with TwistSubscriber and TwistPublisher. I also modified the internal timestamping for the last command packet to use the incoming packet timestamp rather than node wall clock.

Comment (Ryanf55) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:66]: 50cc29245aafb83b29436de78060caec62170afb

File: nav2_util/README.md
Comment (Ryanf55) [reply nav2_util/README.md]: Done c9b4e9b271d0afdda09c7c9b307687dfcbb54dc6

**Comment by Ryanf55** [comment]
> All of the open requests from the last review are also still relevant (but all small). I'll handle the re-pointer-ification once everything else is handled.
> 
> The least of it, but CI is complaining about a small number of linting issues. Usual bits on configuration guide for new parameter + migration guide for capability documentation

Thanks. Almost ready. 

Here's the resources on publishing unique_ptr I tried following. Sadly, none of them led me to a resolution with std::move.
https://design.ros2.org/articles/intraprocess_communications.html
image
https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/

File: nav2_util/include/nav2_util/twist_publisher.hpp
Comment (Ryanf55) [reply nav2_util/include/nav2_util/twist_publisher.hpp:71]: CI is passing on this too.

**Comment by SteveMacenski** [comment]
@Ryanf55 I will follow up in a few moments with a PR to supersede this one with some changes. Please review :-)

**Comment by Ryanf55** [comment]
Feel free to re-open this one. I just pushed a fix that @srmainwaring uncovered during testing.

**Pull request closed without merging**
